src/passes/pass.cpp

```diff
@@ -433,19 +452,31 @@ void PassRunner::addDefaultFunctionOptimizationPasses() {
 }
 
 void PassRunner::addDefaultGlobalOptimizationPrePasses() {
-  add("duplicate-function-elimination");
+  // FIXME DWARF updating does not handle merging debug info with merged code.
+  if (!shouldPreserveDWARF(options, *wasm)) {
+    add("duplicate-function-elimination");
+  }
   add("memory-packing");
 }
 
 void PassRunner::addDefaultGlobalOptimizationPostPasses() {
-  if (options.optimizeLevel >= 2 || options.shrinkLevel >= 1) {
+  auto preserveDWARF = shouldPreserveDWARF(options, *wasm);
+  // FIXME DWARF may be badly affected currently as DAE changes function
+  // signatures and hence params and locals.
+  if (!preserveDWARF &&
+      (options.optimizeLevel >= 2 || options.shrinkLevel >= 1)) {
     add("dae-optimizing");
   }
-  if (options.optimizeLevel >= 2 || options.shrinkLevel >= 2) {
+  // FIXME DWARF updating does not handle inlining yet.
+  if (!preserveDWARF &&
+      (options.optimizeLevel >= 2 || options.shrinkLevel >= 2)) {
```

Comment:
do we need to do something special for inline-main?

Reply:
We disable it on emscripten, https://github.com/emscripten-core/emscripten/pull/10378